**4. What is the effect of using or not using stemming or lemmatization in the precision of n-gram taggers (Question 1 of Week 1)?**

In [1]:
import nltk
nltk.download('punkt') # we need punctuation specs in order to tokenize
nltk.download('averaged_perceptron_tagger') # and the tagger, too (just do these once per computer)
from tokenize import tokenize
from collections import Counter
from nltk.util import ngrams
from nltk import word_tokenize

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.


In [2]:
!pip install gutenbergpy # do not redo this if on your own computer, we already used it in Session 1

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Created wheel for httpsproxy-urllib2: filename=httpsproxy_urllib2-1.0-py3-none-any.whl size=29269 sha256=f5cc1d84a54cfac19cdc954ee41ce9dc336a11017d5c96a97518e7d901277191
  Stored in directory: /root/.cache/pip/wheels/91/04/37/b7080e0cfe33a8bc79ff0082159cea767f9145a5a713b501a3
Successfully built httpsproxy-urllib2


In [3]:
import gutenbergpy.textget # let's reuse what we did in Session 1 to get some example text
target = 105 # this one is Persuasion by Jane Austin

raw  = gutenbergpy.textget.get_text_by_id(target) # content
text = gutenbergpy.textget.strip_headers(raw) # remove header
s = text.decode("utf-8") # get a string from the byte sequence

content = s[s.rindex('Sir Walter Elliot,'):] # using the LAST time the string appears since it is rindex
import re 
noroman = re.sub(r"\bM{0,4}(CM|CD|D?C{0,3})(XC|XL|L?X{0,3})(IX|IV|V?I{0,3})\b", '', content)
clean = re.compile(r'\s+') 
ok = clean.sub(' ', noroman)
potential = ok.split('CHAPTER .')
stripped = [ candidate.strip().lstrip() for candidate in potential ] # leading and trailing space 
chapter = [ s for s in stripped if len(s) > 0 ] # keep only the ones with content
print(len(chapter), ' chapters')
print(chapter[23][:100])

24  chapters
Who can be in doubt of what followed? When any two young people take it into their heads to marry, t


In [ ]:
# Take the first sentence of the chapter 1
text = chapter[0][:103]
text

'Sir Walter Elliot, who united these gifts, was the constant object of his warmest respect and devotion.'

In [ ]:
# This is my gold (pos tag in my sentence)
ref = nltk.pos_tag(word_tokenize(text))
print(ref)

[('Sir', 'NNP'), ('Walter', 'NNP'), ('Elliot', 'NNP'), (',', ','), ('who', 'WP'), ('united', 'VBD'), ('these', 'DT'), ('gifts', 'NNS'), (',', ','), ('was', 'VBD'), ('the', 'DT'), ('constant', 'JJ'), ('object', 'NN'), ('of', 'IN'), ('his', 'PRP$'), ('warmest', 'JJS'), ('respect', 'NN'), ('and', 'CC'), ('devotion', 'NN'), ('.', '.')]


In [ ]:
# take the tag
ref_tag = [i[1] for i in ref]
print(ref_tag)

['NNP', 'NNP', 'NNP', ',', 'WP', 'VBD', 'DT', 'NNS', ',', 'VBD', 'DT', 'JJ', 'NN', 'IN', 'PRP$', 'JJS', 'NN', 'CC', 'NN', '.']


In [ ]:
five_gram = ngrams(sequence=nltk.word_tokenize(text), n=5)

In [ ]:
b=list()
for grams in five_gram:
  t=list(grams)
  b.append(list(grams))
print(b)

[['Sir', 'Walter', 'Elliot', ',', 'who'], ['Walter', 'Elliot', ',', 'who', 'united'], ['Elliot', ',', 'who', 'united', 'these'], [',', 'who', 'united', 'these', 'gifts'], ['who', 'united', 'these', 'gifts', ','], ['united', 'these', 'gifts', ',', 'was'], ['these', 'gifts', ',', 'was', 'the'], ['gifts', ',', 'was', 'the', 'constant'], [',', 'was', 'the', 'constant', 'object'], ['was', 'the', 'constant', 'object', 'of'], ['the', 'constant', 'object', 'of', 'his'], ['constant', 'object', 'of', 'his', 'warmest'], ['object', 'of', 'his', 'warmest', 'respect'], ['of', 'his', 'warmest', 'respect', 'and'], ['his', 'warmest', 'respect', 'and', 'devotion'], ['warmest', 'respect', 'and', 'devotion', '.']]


In [ ]:
pos_test=[]
for i in range(0, len(b)):
  pos_test.append(list(nltk.pos_tag(b[i])))
print(pos_test)

[[('Sir', 'NNP'), ('Walter', 'NNP'), ('Elliot', 'NNP'), (',', ','), ('who', 'WP')], [('Walter', 'NNP'), ('Elliot', 'NNP'), (',', ','), ('who', 'WP'), ('united', 'VBD')], [('Elliot', 'NNP'), (',', ','), ('who', 'WP'), ('united', 'VBD'), ('these', 'DT')], [(',', ','), ('who', 'WP'), ('united', 'VBD'), ('these', 'DT'), ('gifts', 'NNS')], [('who', 'WP'), ('united', 'VBD'), ('these', 'DT'), ('gifts', 'NNS'), (',', ',')], [('united', 'JJ'), ('these', 'DT'), ('gifts', 'NNS'), (',', ','), ('was', 'VBD')], [('these', 'DT'), ('gifts', 'NNS'), (',', ','), ('was', 'VBD'), ('the', 'DT')], [('gifts', 'NNS'), (',', ','), ('was', 'VBD'), ('the', 'DT'), ('constant', 'JJ')], [(',', ','), ('was', 'VBD'), ('the', 'DT'), ('constant', 'JJ'), ('object', 'NN')], [('was', 'VBD'), ('the', 'DT'), ('constant', 'JJ'), ('object', 'NN'), ('of', 'IN')], [('the', 'DT'), ('constant', 'JJ'), ('object', 'NN'), ('of', 'IN'), ('his', 'PRP$')], [('constant', 'JJ'), ('object', 'NN'), ('of', 'IN'), ('his', 'PRP$'), ('warmest'

In [ ]:
five_tag = [[item[1] for item in pos_test] for pos_test in pos_test]
print(five_tag)

[['NNP', 'NNP', 'NNP', ',', 'WP'], ['NNP', 'NNP', ',', 'WP', 'VBD'], ['NNP', ',', 'WP', 'VBD', 'DT'], [',', 'WP', 'VBD', 'DT', 'NNS'], ['WP', 'VBD', 'DT', 'NNS', ','], ['JJ', 'DT', 'NNS', ',', 'VBD'], ['DT', 'NNS', ',', 'VBD', 'DT'], ['NNS', ',', 'VBD', 'DT', 'JJ'], [',', 'VBD', 'DT', 'JJ', 'NN'], ['VBD', 'DT', 'JJ', 'NN', 'IN'], ['DT', 'JJ', 'NN', 'IN', 'PRP$'], ['JJ', 'NN', 'IN', 'PRP$', 'JJS'], ['NN', 'IN', 'PRP$', 'JJS', 'NN'], ['IN', 'PRP$', 'JJS', 'NN', 'CC'], ['PRP$', 'JJS', 'NN', 'CC', 'NN'], ['JJS', 'NN', 'CC', 'NN', '.']]


In [ ]:
a = five_tag[0]
print(a)

import numpy as np
na = np.array(five_tag)
b= na[:,4]
b = b.tolist()
del b[0]
print(b)

five_tags = a+b
print(five_tags)

['NNP', 'NNP', 'NNP', ',', 'WP']
['VBD', 'DT', 'NNS', ',', 'VBD', 'DT', 'JJ', 'NN', 'IN', 'PRP$', 'JJS', 'NN', 'CC', 'NN', '.']
['NNP', 'NNP', 'NNP', ',', 'WP', 'VBD', 'DT', 'NNS', ',', 'VBD', 'DT', 'JJ', 'NN', 'IN', 'PRP$', 'JJS', 'NN', 'CC', 'NN', '.']


In [ ]:
# Comparaison de référence tag, one-gram & five-grams
print('tag_ref: ' + str(ref_tag))
print('tag_five: ' + str(five_tags))

tag_ref: ['NNP', 'NNP', 'NNP', ',', 'WP', 'VBD', 'DT', 'NNS', ',', 'VBD', 'DT', 'JJ', 'NN', 'IN', 'PRP$', 'JJS', 'NN', 'CC', 'NN', '.']
tag_five: ['NNP', 'NNP', 'NNP', ',', 'WP', 'VBD', 'DT', 'NNS', ',', 'VBD', 'DT', 'JJ', 'NN', 'IN', 'PRP$', 'JJS', 'NN', 'CC', 'NN', '.']


In [ ]:
from nltk.metrics import ConfusionMatrix
import collections

ref  = 'NNP NNP NNP WP VBD DT NNS VBD DT JJ NN IN PRP$ JJS NN CC NN'.split()
five = 'NNP NNP NNP WP VBD DT NNS VBD DT JJ NN IN PRP$ JS NN CC NN'.split()
cm = ConfusionMatrix(ref, five)
print(cm)

     |                   P     |
     |         J     N N R V   |
     | C D I J J J N N N P B W |
     | C T N J S S N P S $ D P |
-----+-------------------------+
  CC |<1>. . . . . . . . . . . |
  DT | .<2>. . . . . . . . . . |
  IN | . .<1>. . . . . . . . . |
  JJ | . . .<1>. . . . . . . . |
 JJS | . . . .<.>1 . . . . . . |
  JS | . . . . .<.>. . . . . . |
  NN | . . . . . .<3>. . . . . |
 NNP | . . . . . . .<3>. . . . |
 NNS | . . . . . . . .<1>. . . |
PRP$ | . . . . . . . . .<1>. . |
 VBD | . . . . . . . . . .<2>. |
  WP | . . . . . . . . . . .<1>|
-----+-------------------------+
(row = reference; col = test)



In [ ]:
# https://stackoverflow.com/questions/23704361/how-to-use-the-confusion-matrix-module-in-nltk
labels = set('CC DT IN JJ JJS JS NN NNP NNS PRP$ VBD WP'.split())

true_positives = collections.Counter()
false_negatives = collections.Counter()
false_positives = collections.Counter()

for i in labels:
    for j in labels:
        if i == j:
            true_positives[i] += cm[i,j]
        else:
            false_negatives[i] += cm[i,j]
            false_positives[j] += cm[i,j]

print ("TP:", sum(true_positives.values()), true_positives)
print ("FN:", sum(false_negatives.values()), false_negatives)
print ("FP:", sum(false_positives.values()), false_positives)

TP: 16 Counter({'NNP': 3, 'NN': 3, 'DT': 2, 'VBD': 2, 'PRP$': 1, 'NNS': 1, 'CC': 1, 'WP': 1, 'IN': 1, 'JJ': 1, 'JS': 0, 'JJS': 0})
FN: 1 Counter({'JJS': 1, 'JS': 0, 'NNP': 0, 'NN': 0, 'DT': 0, 'PRP$': 0, 'NNS': 0, 'CC': 0, 'VBD': 0, 'WP': 0, 'IN': 0, 'JJ': 0})
FP: 1 Counter({'JS': 1, 'NNP': 0, 'JJS': 0, 'NN': 0, 'DT': 0, 'PRP$': 0, 'NNS': 0, 'CC': 0, 'VBD': 0, 'WP': 0, 'IN': 0, 'JJ': 0})


In [ ]:
for i in sorted(labels):
    if true_positives[i] == 0:
        fscore = 0
    else:
        precision = true_positives[i] / float(true_positives[i]+false_positives[i])
        recall = true_positives[i] / float(true_positives[i]+false_negatives[i])
        fscore = 2 * (precision * recall) / float(precision + recall)
    print (i, precision)

CC 1.0
DT 1.0
IN 1.0
JJ 1.0
JJS 1.0
JS 1.0
NN 1.0
NNP 1.0
NNS 1.0
PRP$ 1.0
VBD 1.0
WP 1.0


In [ ]:
text

'Sir Walter Elliot, who united these gifts, was the constant object of his warmest respect and devotion.'

In [ ]:
stemmers = nltk.PorterStemmer()
tokens = nltk.word_tokenize(text)

text=[]
for token in tokens: 
  text.append(stemmers.stem(token))

text = ' '.join(text)
text

'sir walter elliot , who unit these gift , wa the constant object of hi warmest respect and devot .'

In [ ]:
five_gram = ngrams(sequence=nltk.word_tokenize(text), n=5)

In [ ]:
b=list()
for grams in five_gram:
  t=list(grams)
  b.append(list(grams))
print(b)

[['sir', 'walter', 'elliot', ',', 'who'], ['walter', 'elliot', ',', 'who', 'unit'], ['elliot', ',', 'who', 'unit', 'these'], [',', 'who', 'unit', 'these', 'gift'], ['who', 'unit', 'these', 'gift', ','], ['unit', 'these', 'gift', ',', 'wa'], ['these', 'gift', ',', 'wa', 'the'], ['gift', ',', 'wa', 'the', 'constant'], [',', 'wa', 'the', 'constant', 'object'], ['wa', 'the', 'constant', 'object', 'of'], ['the', 'constant', 'object', 'of', 'hi'], ['constant', 'object', 'of', 'hi', 'warmest'], ['object', 'of', 'hi', 'warmest', 'respect'], ['of', 'hi', 'warmest', 'respect', 'and'], ['hi', 'warmest', 'respect', 'and', 'devot'], ['warmest', 'respect', 'and', 'devot', '.']]


In [ ]:
pos_test=[]
for i in range(0, len(b)):
  pos_test.append(list(nltk.pos_tag(b[i])))
print(pos_test)

[[('sir', 'NN'), ('walter', 'NN'), ('elliot', 'NN'), (',', ','), ('who', 'WP')], [('walter', 'NN'), ('elliot', 'NN'), (',', ','), ('who', 'WP'), ('unit', 'NN')], [('elliot', 'NN'), (',', ','), ('who', 'WP'), ('unit', 'NN'), ('these', 'DT')], [(',', ','), ('who', 'WP'), ('unit', 'NN'), ('these', 'DT'), ('gift', 'NN')], [('who', 'WP'), ('unit', 'NN'), ('these', 'DT'), ('gift', 'NN'), (',', ',')], [('unit', 'NN'), ('these', 'DT'), ('gift', 'NN'), (',', ','), ('wa', 'NN')], [('these', 'DT'), ('gift', 'NN'), (',', ','), ('wa', 'VBZ'), ('the', 'DT')], [('gift', 'NN'), (',', ','), ('wa', 'VBP'), ('the', 'DT'), ('constant', 'NN')], [(',', ','), ('wa', 'VBZ'), ('the', 'DT'), ('constant', 'JJ'), ('object', 'NN')], [('wa', 'IN'), ('the', 'DT'), ('constant', 'JJ'), ('object', 'NN'), ('of', 'IN')], [('the', 'DT'), ('constant', 'JJ'), ('object', 'NN'), ('of', 'IN'), ('hi', 'NN')], [('constant', 'JJ'), ('object', 'NN'), ('of', 'IN'), ('hi', 'NN'), ('warmest', 'JJS')], [('object', 'NN'), ('of', 'IN'),

In [ ]:
five_tag = [[item[1] for item in pos_test] for pos_test in pos_test]
print(five_tag)

[['NN', 'NN', 'NN', ',', 'WP'], ['NN', 'NN', ',', 'WP', 'NN'], ['NN', ',', 'WP', 'NN', 'DT'], [',', 'WP', 'NN', 'DT', 'NN'], ['WP', 'NN', 'DT', 'NN', ','], ['NN', 'DT', 'NN', ',', 'NN'], ['DT', 'NN', ',', 'VBZ', 'DT'], ['NN', ',', 'VBP', 'DT', 'NN'], [',', 'VBZ', 'DT', 'JJ', 'NN'], ['IN', 'DT', 'JJ', 'NN', 'IN'], ['DT', 'JJ', 'NN', 'IN', 'NN'], ['JJ', 'NN', 'IN', 'NN', 'JJS'], ['NN', 'IN', 'NN', 'JJS', 'NN'], ['IN', 'JJ', 'JJS', 'NN', 'CC'], ['NN', 'JJS', 'NN', 'CC', 'NN'], ['JJS', 'NN', 'CC', 'NN', '.']]


In [ ]:
a = five_tag[0]
print(a)

import numpy as np
na = np.array(five_tag)
b= na[:,4]
b = b.tolist()
del b[0]
print(b)

five_tags = a+b
print(five_tags)

['NN', 'NN', 'NN', ',', 'WP']
['NN', 'DT', 'NN', ',', 'NN', 'DT', 'NN', 'NN', 'IN', 'NN', 'JJS', 'NN', 'CC', 'NN', '.']
['NN', 'NN', 'NN', ',', 'WP', 'NN', 'DT', 'NN', ',', 'NN', 'DT', 'NN', 'NN', 'IN', 'NN', 'JJS', 'NN', 'CC', 'NN', '.']


In [ ]:
# Comparaison de référence tag, one-gram & five-grams
print('tag_ref: ' + str(ref_tag))
print('tag_five: ' + str(five_tags))

tag_ref: ['NNP', 'NNP', 'NNP', ',', 'WP', 'VBD', 'DT', 'NNS', ',', 'VBD', 'DT', 'JJ', 'NN', 'IN', 'PRP$', 'JJS', 'NN', 'CC', 'NN', '.']
tag_five: ['NN', 'NN', 'NN', ',', 'WP', 'NN', 'DT', 'NN', ',', 'NN', 'DT', 'NN', 'NN', 'IN', 'NN', 'JJS', 'NN', 'CC', 'NN', '.']


In [ ]:
from nltk.metrics import ConfusionMatrix
import collections

ref  = 'NNP NNP NNP WP VBD DT NNS VBD DT JJ NN IN PRP$ JJS NN CC NN'.split()
five = 'NN NN NN WP NN DT NN NN DT NN NN IN NN JJS NN CC NN'.split()
cm = ConfusionMatrix(ref, five)
print(cm)

     |                 P     |
     |         J   N N R V   |
     | C D I J J N N N P B W |
     | C T N J S N P S $ D P |
-----+-----------------------+
  CC |<1>. . . . . . . . . . |
  DT | .<2>. . . . . . . . . |
  IN | . .<1>. . . . . . . . |
  JJ | . . .<.>. 1 . . . . . |
 JJS | . . . .<1>. . . . . . |
  NN | . . . . .<3>. . . . . |
 NNP | . . . . . 3<.>. . . . |
 NNS | . . . . . 1 .<.>. . . |
PRP$ | . . . . . 1 . .<.>. . |
 VBD | . . . . . 2 . . .<.>. |
  WP | . . . . . . . . . .<1>|
-----+-----------------------+
(row = reference; col = test)



In [ ]:
# https://stackoverflow.com/questions/23704361/how-to-use-the-confusion-matrix-module-in-nltk
labels = set('CC DT IN JJ JJS NN NNP NNS PRP$ VBD WP'.split())

true_positives = collections.Counter()
false_negatives = collections.Counter()
false_positives = collections.Counter()

for i in labels:
    for j in labels:
        if i == j:
            true_positives[i] += cm[i,j]
        else:
            false_negatives[i] += cm[i,j]
            false_positives[j] += cm[i,j]

print ("TP:", sum(true_positives.values()), true_positives)
print ("FN:", sum(false_negatives.values()), false_negatives)
print ("FP:", sum(false_positives.values()), false_positives)

TP: 9 Counter({'NN': 3, 'DT': 2, 'JJS': 1, 'CC': 1, 'WP': 1, 'IN': 1, 'NNP': 0, 'PRP$': 0, 'NNS': 0, 'VBD': 0, 'JJ': 0})
FN: 8 Counter({'NNP': 3, 'VBD': 2, 'PRP$': 1, 'NNS': 1, 'JJ': 1, 'JJS': 0, 'NN': 0, 'DT': 0, 'CC': 0, 'WP': 0, 'IN': 0})
FP: 8 Counter({'NN': 8, 'JJS': 0, 'DT': 0, 'PRP$': 0, 'NNS': 0, 'CC': 0, 'VBD': 0, 'WP': 0, 'IN': 0, 'JJ': 0, 'NNP': 0})


In [ ]:
for i in sorted(labels):
    if true_positives[i] == 0:
        fscore = 0
    else:
        precision = true_positives[i] / float(true_positives[i]+false_positives[i])
        recall = true_positives[i] / float(true_positives[i]+false_negatives[i])
        fscore = 2 * (precision * recall) / float(precision + recall)
    print (i, precision)

CC 1.0
DT 1.0
IN 1.0
JJ 1.0
JJS 1.0
NN 0.2727272727272727
NNP 0.2727272727272727
NNS 0.2727272727272727
PRP$ 0.2727272727272727
VBD 0.2727272727272727
WP 1.0


La précision diminue avec le stemming effectué sur un 5-grams. En particulier sur les noms, pronom et verbe selon ma phrase sélectionné. À noter qu’ici je n’ai pas supprimé les stop words en raison du type d’analyse pos tag. Le pos tag devrait être utilisé avant le stemming par conséquent.

**5. How are the transformer-based taggers affected by stemming or lemmatization (Question 3 of Week 4)?**

In [ ]:
pip install flair

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 401 kB 5.1 MB/s 
     |████████████████████████████████| 163 kB 58.8 MB/s 
     |████████████████████████████████| 46 kB 3.5 MB/s 
     |████████████████████████████████| 53 kB 1.7 MB/s 
     |████████████████████████████████| 19.7 MB 4.3 MB/s 
     |████████████████████████████████| 5.3 MB 34.1 MB/s 
     |████████████████████████████████| 788 kB 51.6 MB/s 
     |████████████████████████████████| 1.2 MB 12.0 MB/s 
     |████████████████████████████████| 981 kB 51.4 MB/s 
     |████████████████████████████████| 1.6 MB 51.8 MB/s 
     |████████████████████████████████| 200 kB 62.1 MB/s 
     |████████████████████████████████| 62 kB 1.3 MB/s 
     |████████████████████████████████| 7.6 MB 44.3 MB/s 
  Created wheel for mpld3: filename=mpld3-0.3-py3-none-any.whl size=116702 sha256=0f740a05a9fefb79e70d008d63c5658a84531e505224806c0eed4030df0676c8
  Store

In [ ]:
import time

In [ ]:
from flair.data import Sentence
from flair.models import SequenceTagger

# load tagger
tagger = SequenceTagger.load("flair/pos-english")

# make example sentence
sentence = Sentence(chapter[0])

# predict NER tags
start = time.time()
tagger.predict(sentence)
stop = time.time()
print(f"Training time: {stop - start}s")

2022-10-19 20:30:26,058 loading file /root/.flair/models/pos-english/a9a73f6cd878edce8a0fa518db76f441f1cc49c2525b2b4557af278ec2f0659e.121306ea62993d04cd1978398b68396931a39eb47754c8a06a87f325ea70ac63


/usr/local/lib/python3.7/dist-packages/huggingface_hub/file_download.py:591: FutureWarning: `cached_download` is the legacy way to download files from the HF hub, please consider upgrading to `hf_hub_download`
  FutureWarning,


2022-10-19 20:30:26,394 SequenceTagger predicts: Dictionary with 53 tags: <unk>, O, UH, ,, VBD, PRP, VB, PRP$, NN, RB, ., DT, JJ, VBP, VBG, IN, CD, NNS, NNP, WRB, VBZ, WDT, CC, TO, MD, VBN, WP, :, RP, EX, JJR, FW, XX, HYPH, POS, RBR, JJS, PDT, NNPS, RBS, AFX, WP$, -LRB-, -RRB-, ``, '', LS, $, SYM, ADD
Training time: 8.480138778686523s


In [ ]:
# iterate over all 'ner'-labels in the sentence
for label in sentence.get_labels('pos'):
    # print label value and score
    print(f'label.value is: "{label.value}"')
    print(f'label.score is: "{label.score}"')
    # access the data point to which label attaches and print its text
    print(f'the text of label.data_point is: "{label.data_point.text}"\n')

Le flux de sortie a été tronqué et ne contient que les 5000 dernières lignes.
label.value is: "RB"
label.score is: "1.0"
the text of label.data_point is: "ever"

label.value is: "VB"
label.score is: "0.9999947547912598"
the text of label.data_point is: "present"

label.value is: "DT"
label.score is: "1.0"
the text of label.data_point is: "the"

label.value is: "NN"
label.score is: "0.9999983310699463"
the text of label.data_point is: "remembrance"

label.value is: "IN"
label.score is: "1.0"
the text of label.data_point is: "of"

label.value is: "."
label.score is: "0.7906712293624878"
the text of label.data_point is: "."

label.value is: "DT"
label.score is: "1.0"
the text of label.data_point is: "The"

label.value is: "NN"
label.score is: "0.9966855645179749"
the text of label.data_point is: "heir"

label.value is: "JJ"
label.score is: "0.9996652603149414"
the text of label.data_point is: "presumptive"

label.value is: ","
label.score is: "1.0"
the text of label.data_point is: ","

la

In [ ]:
stemmers = nltk.PorterStemmer()
tokens = nltk.word_tokenize(chapter[0])

text=[]
for token in tokens: 
  text.append(stemmers.stem(token))

chapter = ' '.join(text)
chapter

'sir walter elliot , who unit these gift , wa the constant object of hi warmest respect and devot . hi good look and hi rank had one fair claim on hi attach ; sinc to them he must have owe a wife of veri superior charact to ani thing deserv by hi own . ladi elliot had been an excel woman , sensibl and amiabl ; whose judgement and conduct , if they might be pardon the youth infatu which made her ladi elliot , had never requir indulg afterwards.—sh had humour , or soften , or conceal hi fail , and promot hi real respect for seventeen year ; and though not the veri happiest be in the world herself , had found enough in her duti , her friend , and her children , to attach her to life , and make it no matter of indiffer to her when she wa call on to quit them.—thre girl , the two eldest sixteen and fourteen , wa an aw legaci for a mother to bequeath , an aw charg rather , to confid to the author and guidanc of a conceit , silli father . she had , howev , one veri intim friend , a sensibl , 

In [ ]:
from flair.data import Sentence
from flair.models import SequenceTagger

# load tagger
tagger = SequenceTagger.load("flair/pos-english")

# make example sentence
sentence = Sentence(chapter)

# predict NER tags
start = time.time()
tagger.predict(sentence)
stop = time.time()
print(f"Training time: {stop - start}s")

2022-10-19 20:31:47,023 loading file /root/.flair/models/pos-english/a9a73f6cd878edce8a0fa518db76f441f1cc49c2525b2b4557af278ec2f0659e.121306ea62993d04cd1978398b68396931a39eb47754c8a06a87f325ea70ac63


/usr/local/lib/python3.7/dist-packages/huggingface_hub/file_download.py:591: FutureWarning: `cached_download` is the legacy way to download files from the HF hub, please consider upgrading to `hf_hub_download`
  FutureWarning,


2022-10-19 20:31:47,281 SequenceTagger predicts: Dictionary with 53 tags: <unk>, O, UH, ,, VBD, PRP, VB, PRP$, NN, RB, ., DT, JJ, VBP, VBG, IN, CD, NNS, NNP, WRB, VBZ, WDT, CC, TO, MD, VBN, WP, :, RP, EX, JJR, FW, XX, HYPH, POS, RBR, JJS, PDT, NNPS, RBS, AFX, WP$, -LRB-, -RRB-, ``, '', LS, $, SYM, ADD
Training time: 7.561844348907471s


In [ ]:
# iterate over all 'ner'-labels in the sentence
for label in sentence.get_labels('pos'):
    # print label value and score
    print(f'label.value is: "{label.value}"')
    print(f'label.score is: "{label.score}"')
    # access the data point to which label attaches and print its text
    print(f'the text of label.data_point is: "{label.data_point.text}"\n')

Le flux de sortie a été tronqué et ne contient que les 5000 dernières lignes.
label.value is: ","
label.score is: "1.0"
the text of label.data_point is: ","

label.value is: "DT"
label.score is: "1.0"
the text of label.data_point is: "the"

label.value is: "JJ"
label.score is: "0.9192838072776794"
the text of label.data_point is: "veri"

label.value is: "NN"
label.score is: "0.9579282999038696"
the text of label.data_point is: "william"

label.value is: "NN"
label.score is: "0.9922886490821838"
the text of label.data_point is: "walter"

label.value is: "NNP"
label.score is: "0.970488429069519"
the text of label.data_point is: "elliot"

label.value is: ","
label.score is: "1.0"
the text of label.data_point is: ","

label.value is: "NNP"
label.score is: "0.6706222295761108"
the text of label.data_point is: "esq"

label.value is: "."
label.score is: "0.980970025062561"
the text of label.data_point is: "."

label.value is: ","
label.score is: "1.0"
the text of label.data_point is: ","

lab

label.value is: "RB"
label.score is: "0.9998332262039185"
the text of label.data_point is: "very"

label.value is: "JJ"
label.score is: "0.8345663547515869"
the text of label.data_point is: "veri"

Même effet avec le transformer. Le mot very est transformé par veri avec le stemming et le score pos tag diminue.


**6. In order to use a similarity quantifier (like the one of Question 2 of Week 5) to detect suspected plagiarism, would you expect this to be more useful for this purpose or less so if lemmatization or stemming were used explicitly before querying for WordNet similarities?**

In [4]:
import nltk # this we already have
nltk.download('wordnet') # this is new, download once per environment
nltk.download('omw-1.4') # same here
from nltk.corpus import wordnet as wn
from nltk.tokenize import word_tokenize
nltk.download('punkt')

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [5]:
sent_1 = 'The roof of my house needs repair.'
sent_2 = 'The pretty birds eat vegetables from my garden.'
sent_3 = 'This lakeside cottage is amazing.'

In [7]:
from nltk.corpus import stopwords
nltk.download('stopwords')

#Stopword
stop_words = set(stopwords.words('english'))

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [8]:
def change_word(sentence):

  sent=[]

  # Tokenize the sentence
  sent_1w = word_tokenize(sentence)

  for i in range(0, len(sent_1w)):

    # check stopwords and change nothing if is in the list of stopwords
    if sent_1w[i] in stop_words:
      sent.append(sent_1w[i])
  
    else:
      synset = wn.synsets(sent_1w[i].lower())

      # check is the word have synset, if not append in the list
      if len(synset) == 0:
        sent.append(sent_1w[i])

      # if the word have synset, take a similar word
      else:
        ss = synset[0] 
        lm = ss.lemma_names()
        lm = lm[0]

        # Choose lemma_names different of the word we want change
        if lm != sent_1w[i]:
          sent.append(lm)
        else:
          ss = synset[-1] 
          lm = ss.lemma_names()
          lm = lm[-1]
          sent.append(lm)

  return sent

In [9]:
# Sentece 1
print('Transform: ' + str(' '.join(change_word(sent_1))))
print('Sentence: ' + str(sent_1))

Transform: The roof of my domiciliate need revivify .
Sentence: The roof of my house needs repair.


In [29]:
# Sentece 2
print('Transform: ' + str(' '.join(change_word(sent_2))))
print('Sentence: ' + str(sent_2))

Transform: The passably bird rust vegetable from my garden .
Sentence: The pretty birds eat vegetables from my garden.


In [11]:
# Sentece 3
print('Transform: ' + str(' '.join(change_word(sent_3))))
print('Sentence: ' + str(sent_3))

Transform: This lakeshore bungalow is amaze .
Sentence: This lakeside cottage is amazing.


In [30]:
stemmers = nltk.PorterStemmer()

tokens_1 = nltk.word_tokenize(sent_1)
tokens_2 = nltk.word_tokenize(sent_2)
tokens_3 = nltk.word_tokenize(sent_3)

tokens_4 = nltk.word_tokenize('The roof of my domiciliate need revivify .')
tokens_5 = nltk.word_tokenize('The passably bird rust vegetable from my garden .')
tokens_6 = nltk.word_tokenize('This lakeshore bungalow is amaze .')

text_1=[]
text_2=[]
text_3=[]
text_4=[]
text_5=[]
text_6=[]

for token in tokens_1: 
  text_1.append(stemmers.stem(token))
for token in tokens_2: 
  text_2.append(stemmers.stem(token))
for token in tokens_3: 
  text_3.append(stemmers.stem(token))
for token in tokens_4: 
  text_4.append(stemmers.stem(token))
for token in tokens_5: 
  text_5.append(stemmers.stem(token))
for token in tokens_6: 
  text_6.append(stemmers.stem(token))

text_1 = ' '.join(text_1)
text_2 = ' '.join(text_2)
text_3 = ' '.join(text_3)
text_4 = ' '.join(text_4)
text_5 = ' '.join(text_5)
text_6 = ' '.join(text_6)

print(text_1)
print(text_2)
print(text_3)
print(text_4)
print(text_5)
print(text_6)

the roof of my hous need repair .
the pretti bird eat veget from my garden .
thi lakesid cottag is amaz .
the roof of my domicili need revivifi .
the passabl bird rust veget from my garden .
thi lakeshor bungalow is amaz .


In [15]:
!python -m spacy download en_core_web_md

2022-10-23 22:20:35.243340: E tensorflow/stream_executor/cuda/cuda_driver.cc:271] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 42.8 MB 2.0 MB/s 
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_md')


In [16]:
import spacy
nlp = spacy.load('en_core_web_md')

In [32]:
sentences_1 = ['The roof of my domiciliate need revivify.', 'The roof of my house needs repair.']
sentences_2 = ['The passably bird are rust my vegetable from my garden.', 'The pretty birds are eating my vegetables from my garden.']
sentences_3 = ['This lakeshore bungalow is amaze.', 'This lakeside cottage is amazing.']

In [31]:
sentences_1s = ['the roof of my domicili need revivifi.', 'the roof of my hous need repair.']
sentences_2s = ['the passabl bird rust veget from my garden.', 'the pretti bird eat veget from my garden.']
sentences_3s = ['thi lakeshor bungalow is amaz.', 'thi lakesid cottag is amaz.']

In [33]:
embeddings_1 = [nlp(sentence).vector for sentence in sentences_1]
embeddings_2 = [nlp(sentence).vector for sentence in sentences_2]
embeddings_3 = [nlp(sentence).vector for sentence in sentences_3]

embeddings_1s = [nlp(sentence).vector for sentence in sentences_1s]
embeddings_2s = [nlp(sentence).vector for sentence in sentences_2s]
embeddings_3s = [nlp(sentence).vector for sentence in sentences_3s]

In [34]:
from math import sqrt, pow, exp
 
def squared_sum(x):
  """ return 3 rounded square rooted value """
 
  return round(sqrt(sum([a*a for a in x])),3)

In [35]:
def cos_similarity(x,y):
  """ return cosine similarity between two lists """
 
  numerator = sum(a*b for a,b in zip(x,y))
  denominator = squared_sum(x)*squared_sum(y)
  return round(numerator/float(denominator),3)

In [36]:
cos_1 = cos_similarity(embeddings_1[0], embeddings_1[1])
cos_2 = cos_similarity(embeddings_2[0], embeddings_2[1])
cos_3 = cos_similarity(embeddings_3[0], embeddings_3[1])
print(cos_1, cos_2, cos_3)

0.956 0.973 0.959


In [37]:
cos_1s = cos_similarity(embeddings_1s[0], embeddings_1s[1])
cos_2s = cos_similarity(embeddings_2s[0], embeddings_2s[1])
cos_3s = cos_similarity(embeddings_3s[0], embeddings_3s[1])
print(cos_1s, cos_2s, cos_3s)

0.98 0.952 0.975


Si je fais la mesure de similarité cosine sur les phrases avec strem, ceci augmente généralement le résultat. 

Sans stemming: 0.956 0.973 0.959

Avec stemming: 0.98 0.952 0.975

Le stemming semble être une bonne stratégie dans ce type d’analyse. Voir ci-dessous deux références auxquels ils utilisent le lemmatization dans le preprocessing pour l’analyse de plagiat.

https://aclanthology.org/R09-1011.pdf	

https://www.researchgate.net/publication/355800945_Plagiarism_Detection_Algorithm_Model_Based_on_NLP_Technology


**7. Would text prediction (like the auto-complete of Question 2 of Week 6) be easier in some way if stemming or lemmatization were included in some part of the process?**

In [ ]:
# Upload the file in Google Colab
from google.colab import files
uploaded = files.upload()

Saving smart-home-commands-dataset.zip to smart-home-commands-dataset.zip


In [ ]:
import pandas as pd
import itertools

df = pd.read_csv('smart-home-commands-dataset.zip')
df.head()

Number Category  Action_needed  Question Subcategory Action      Time  \
0       1   lights              1         0     kitchen     on     today   
1       2   lights              1         0     kitchen     on  tomorrow   
2       3   lights              1         0     kitchen     on      hour   
3       4   lights              1         0     kitchen     on       day   
4       5   lights              1         0  diningroom     on     today   

                                        Sentence  
0                  Illuminate the kitchen today.  
1               Illuminate the kitchen tomorrow.  
2  Turn on the light in the kitchen in 10 hours.  
3     Turn on the light in the kitchen in 1 day.  
4              Illuminate the dining room today.

In [ ]:
df_sent = df[['Sentence']]
sent = [word_tokenize(sentence['Sentence']) for index, sentence in df_sent.iterrows()]
merge_sent = list(itertools.chain.from_iterable(sent))

In [ ]:
from heapq import nlargest
from operator import itemgetter

In [ ]:
def auto_complete_tri(sentence, n):

  find = nltk.TrigramCollocationFinder.from_words(merge_sent) 
  pmi = find.score_ngrams(nltk.TrigramAssocMeasures().pmi)

  beg_sentence = nltk.word_tokenize(sentence)
  w1 = beg_sentence[0]
  w2 = beg_sentence[1]

  top=[]
  for (tri, score) in pmi:
    (first, second, third) = tri

    if first == w1 and second == w2:
      top.append((third, score))
    
  top_max = nlargest(n, top, key=itemgetter(1))

  return (top_max)

In [ ]:
auto_complete_tri('Illuminate the', 5)

[('dining', 9.821787027232098),
 ('kitchen', 9.049197523335174),
 ('attic', 8.59939460589565)]

In [ ]:
text_merge = ' '.join(merge_sent)

In [ ]:
stemmers = nltk.PorterStemmer()
tokens = nltk.word_tokenize(text_merge)

text=[]
for token in tokens: 
  text.append(stemmers.stem(token))

text = ' '.join(text)
print(text)

illumin the kitchen today . illumin the kitchen tomorrow . turn on the light in the kitchen in 10 hour . turn on the light in the kitchen in 1 day . illumin the dine room today . do n't illumin the dine room today . illumin the dine room tomorrow . turn on the light in the dinin room in 10 hour . turn on the light in the dine room in 1 day . turn on the light in the bathroom . do n't turn on the light right now . i would get mad if you 'd put on the light right now . i would not like it if you would n't turn on the light right now . is the light on in the live room . would you illumin the live room in one hour for me pleas . it would be great if you 'd turn on the light for me . light on in kitchen . light off in live room in a few hour . turn the light off in the basement . turn the light off in the dine room . i would like if you made it brighter . i would like if you made it brighter in 3 minut . can you make it brighter here in 8 minut ? can you make it more dark here ? do n't make

In [ ]:
merge_sent = word_tokenize(text) 
#print(sent)

In [ ]:
stemmers = nltk.PorterStemmer()
tokens = nltk.word_tokenize('Illuminate the')
text=[]
for token in tokens: 
  text.append(stemmers.stem(token))
text = ' '.join(text)
text

'illumin the'

In [ ]:
auto_complete_tri('illumin the', 5)

[('dine', 9.857714317857397),
 ('live', 8.5946799120236),
 ('kitchen', 8.559056002292877),
 ('attic', 8.05862701178339)]

auto_complete_tri('Illuminate the', 5)

[('dining', 9.821787027232098), ('kitchen', 9.049197523335174), ('attic', 8.59939460589565)]

[('dine', 9.857714317857397), ('live', 8.5946799120236), ('kitchen', 8.559056002292877), ('attic', 8.05862701178339)]

En appliquant le stemming, l’auto-complete trouve plus de mots à suggérer. (ex : apple et apples seront considéré comme un même mot). Il est donc avantageux d’utiliser cette technique pour ce type d’analyse.

https://github.com/ankane/autosuggest

https://towardsdatascience.com/github-autocompletion-with-machine-learning-a833cb90983e
